<!DOCTYPE html>
<html lang="en">

  <!-- Head -->
  <head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">    <!-- Metadata, OpenGraph and Schema.org -->
    

<!-- Standard metadata -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>talks | Michael  Bleher</title>
    <meta name="author" content="Michael  Bleher">
    <meta name="description" content="personal website
">
    <meta name="keywords" content="mathematical physics, physical mathematics, topological data analysis, computational topology, geometric machine learning">


    <!-- Bootstrap & MDB -->
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04">
    <!-- <link rel="stylesheet" href="/assets/css/mdb.min.css?62a43d1430ddb46fc4886f9d0e3b49b8"> -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous">

    <!-- Bootstrap Table -->
    <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css">

    <!-- Fonts & Icons -->
    <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5">
    <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons">

    <!-- Code Syntax Highlighting -->
    <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light">

    
    <!-- Sidebar Table of Contents -->
    <link href="/assets/css/bootstrap-toc.min.css?6f5af0bb9aab25d79b2448143cbeaa88" rel="stylesheet">
    

    

    <!-- Styles -->
    
    <link rel="shortcut icon" href="/assets/img/favicon.png?2689c4097973519343e51c901211cdd5">
    
    <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e">
    <link rel="canonical" href="https://subthaumic.github.io/talks/">

    <!-- Dark Mode -->
    
    <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark">
    <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script>
    <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script>
    

  </head>

  <!-- Body -->
  <body class="fixed-top-nav sticky-bottom-footer">

    <!-- Header --><header>

  <!-- Nav Bar -->
  <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top">
    <div class="container">
      <a class="navbar-brand title font-weight-lighter" href="/"><span class="font-weight-bold">MichaelÂ </span>Bleher</a>
      <!-- Navbar Toggle -->
      <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar top-bar"></span>
        <span class="icon-bar middle-bar"></span>
        <span class="icon-bar bottom-bar"></span>
      </button>

      <div class="collapse navbar-collapse text-right" id="navbarNav">
        <ul class="navbar-nav ml-auto flex-nowrap">

          <!-- About -->
          <li class="nav-item ">
            <a class="nav-link" href="/">about</a>
          </li>

          <!-- Other pages -->
            <li class="nav-item "><a class="nav-link" href="/projects/">projects</a></li>
            <li class="nav-item "><a class="nav-link" href="/repos/">repos</a></li>
            <li class="nav-item "><a class="nav-link" href="/publications/">publications</a></li>
            <li class="nav-item active"><a class="nav-link" href="/talks/">talks<span class="sr-only">(current)</span></a></li>
            <li class="nav-item "><a class="nav-link" href="/teaching/">teaching</a></li>
            <li class="nav-item "><a class="nav-link" href="https://subthaumic.substack.com/" rel="external nofollow noopener" target="_blank">blog</a></li>
            <li class="nav-item "><a class="nav-link" href="/cv/">CV</a></li>
            <!-- Toogle theme mode -->
            <li class="toggle-container">
              <button id="light-toggle" title="Change theme">
                <i class="fa-solid fa-moon"></i>
                <i class="fa-solid fa-sun"></i>
              </button>
            </li>
        </ul>
      </div>
    </div>
  </nav>

  
      <!-- Scrolling Progress Bar -->
      <progress id="progress" value="0">
        <div class="progress-container">
          <span class="progress-bar"></span>
        </div>
      </progress>
</header>


    <!-- Content -->
    <div class="container mt-5">
      
        
        <div class="row">
          <!-- sidebar, which will move to the top on a small screen -->
          <div class="col-sm-3">
            <nav id="toc-sidebar" class="sticky-top"></nav>
          </div>
          <!-- main content area -->
          <div class="col-sm-9">
            <!-- page.html -->
<div class="post">

  <header class="post-header">
    <h1 class="post-title">talks</h1>
    <p class="post-description"></p>
</header>

  <article>
    <!-- _pages/talks.md -->
<div class="talks">

  <!-- Get talks and sort them by date -->

  <!-- Extract unique years from the talks -->

  <!-- Display talks by year -->
  
    
      <h2 class="talks"> 2025</h2>

      <!-- Generate cards for each talk -->
      <div class="container">
      <div class="row row-cols-1">
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">The tangled web they weave -- Exploring neural networks with directed topology</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>Young Topologists Meeting 2025, Stockholm</strong> <br>
              <i>June 25, 2025</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2025_ytm.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse --><div class="abstract">
                <a href="#" class="expand-link" onclick="toggleAbstract('abstract-1'); return false;">
                  Abstract
                </a>
                <p id="abstract-1" class="card-text abstract-text" style="display:none;">
                  We propose a methodology to construct filtered directed simplicial complexes from the internal states of artificial neural networks (ANNs). Our approach treats ANNs as structural causal models and applies Pearl's counterfactual reasoning to infer directed relationships between neuron activations in a given input context. We argue that the resulting complexes provide a meaningful model of distributed feature representations and offer insights into their superposition at polysemantic neurons. As an initial example, we analyse activation data from a multilayer perceptron trained on a simple classification task, demonstrating how directed simplices capture the network's capacity to isolate and recombine distinct features. This framework offers a principled approach to characterising neural representations, with potential applications in model evaluation and feature disentanglement.
                </p>
              </div>
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">Harmonic Persistent Homology</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>Persistent Seminar, University of Heidelberg</strong> <br>
              <i>May 28, 2025</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2025_persistent-seminar.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse -->
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
        
        
        
        
        
        
      </div>
      </div>
    
  
    
      <h2 class="talks"> 2024</h2>

      <!-- Generate cards for each talk -->
      <div class="container">
      <div class="row row-cols-1">
        
        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">Persistence and Coarse-Graining in Dynamical Biological Systems</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>TDA Symposium, MPI for molecular cell biology and genomics, Dresden</strong> <br>
              <i>December 12, 2024</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2024_dresden.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse -->
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">Fast computation of pathwise persistence in pandemic scale SARS-CoV-2 genome data</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>4th workshop on Computational Persistence, Graz</strong> <br>
              <i>September 24, 2024</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2024_graz.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse --><div class="abstract">
                <a href="#" class="expand-link" onclick="toggleAbstract('abstract-4'); return false;">
                  Abstract
                </a>
                <p id="abstract-4" class="card-text abstract-text" style="display:none;">
                  This talk showcases a methodology for the rapid computation of persistent homology in pandemic-scale time series data of SARS-CoV-2 genomes. By focusing only on small genomic distance scales, which capture the selective advantages and deleterious effects of individual mutations relative to its progenitor, and encoding temporal information through a deformation of the metric, we are able to efficiently calculate homology classes for millions of unique viral sequences. We discuss the implementation challenges and solutions that have enabled us to process data at this scale and explain the biological insights we obtain about the evolutionary fitness landscape during the pandemic. Our method exemplifies the application of topology to large datasets, streamlining data analysis at unprecedented scales, while also enhancing our understanding of viral evolution by providing insights into mutation dynamics at a daily resolution.
                </p>
              </div>
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">Haydys-Witten instantons and the gauge theoretic approach to Khovanov Homology</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>Gauge Theory and Mathematical Physics Seminar, Morningside Center of Mathematics, Beijing</strong> <br>
              <i>July 03, 2024</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2024_beijing.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse --><div class="abstract">
                <a href="#" class="expand-link" onclick="toggleAbstract('abstract-5'); return false;">
                  Abstract
                </a>
                <p id="abstract-5" class="card-text abstract-text" style="display:none;">
                  The Haydys-Witten equations are gauge theoretic partial differential equations on five-dimensional Riemannian manifolds that are equipped with a non-vanishing vector field. Conjecturally, their solutions on M^5=IRÃW^4 determine the Floer differential in a gauge-theoretic approach to homological knot invariants. In this talk, I will provide a brief overview of this âinstanton Floer theory of four-manifoldsâ and then focus on a âdecoupledâ version of the Haydys-Witten equations that emerges in the geometric setup for knot invariants. Since the latter exhibit a Hermitian Yang-Mills structure, these results may offer novel insights into the conjectured relationship to knot invariants.
                </p>
              </div>
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">RNA velocity embeddings in curved spaces -- Exploring cellular dynamics</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>Dagstuhl Seminar 24122</strong> <br>
              <i>March 20, 2024</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2024_dagstuhl.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse --><div class="abstract">
                <a href="#" class="expand-link" onclick="toggleAbstract('abstract-6'); return false;">
                  Abstract
                </a>
                <p id="abstract-6" class="card-text abstract-text" style="display:none;">
                  RNA velocity data provides a snapshot of cell states and their current rate of change. It promises insights into the behaviour of individual cells and the dynamics governing cell division and differentiation processes. To explore single cell RNA-sequence data one often relies on low-dimensional visualizations, e.g. tSNE or UMAP. A priori it's not obvious how RNA velocities carry over to such representations and current methods have several drawbacks. It was recently suggested that one should fix a biologically motivated, low-dimensional manifold and infer RNA velocities strictly in terms of an embedding of the data in that manifold. I expand on that idea and argue that low-dimensional representations of position-velocity pairs should utilize the Sasakian geometry on the tangent bundle of curved target spaces. Moreover, I propose that non-linear neighbour embeddings into low- or middle-dimensional symmetric spaces provide a geometric representation of the principal dynamical components in the data. This geometrization provides interesting future directions regarding the analysis of the dynamical processes captured in single cell data.
                </p>
              </div>
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">On Haydys-Witten instantons and the gauge theoretic approach to Khovanov Homology</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>HU Gauge Theory Research Seminar, Berlin</strong> <br>
              <i>January 31, 2024</i>
            </p>

            <!-- Optional link to slides --><!-- Optional abstract with expand/collapse --><div class="abstract">
                <a href="#" class="expand-link" onclick="toggleAbstract('abstract-7'); return false;">
                  Abstract
                </a>
                <p id="abstract-7" class="card-text abstract-text" style="display:none;">
                  The Haydys-Witten equations are partial differential equations on five-dimensional Riemannian manifolds that are equipped with a non-vanishing vector field v. Conjecturally, their solutions on M5=RÃW4 determine the Floer differential in a gauge-theoretic approach to homological knot invariants. In this talk, I will provide a brief overview of this âinstanton Floer theory of four-manifoldsâ and then focus on a âdecoupledâ version of the Haydys-Witten equations that emerges in the geometric setup for knot invariants. Since the latter exhibit a Hermitian Yang-Mills structure, these results may offer novel insights into the conjectured relationship to knot invariants
                </p>
              </div>
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
        
      </div>
      </div>
    
  
    
      <h2 class="talks"> 2023</h2>

      <!-- Generate cards for each talk -->
      <div class="container">
      <div class="row row-cols-1">
        
        
        
        
        
        
        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">Haydys-Witten instantons in the gauge theoretic approach to Khovanov Homology</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>ULB geometry seminar, Brussels</strong> <br>
              <i>December 04, 2023</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2023_brussels.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse --><div class="abstract">
                <a href="#" class="expand-link" onclick="toggleAbstract('abstract-8'); return false;">
                  Abstract
                </a>
                <p id="abstract-8" class="card-text abstract-text" style="display:none;">
                  The Haydys-Witten equations are partial differential equations in gauge theory, defined on five-dimensional Riemannian manifolds that are equipped with a non-vanishing vector field $v$. Conjecturally, their solutions on $M^5 = \mathbb{R} \times W^4$ determine the Floer differential in a gauge-theoretic approach to homological knot invariants. In this talk, I will provide a brief overview of this 'instanton Floer theory of four-manifolds' and then focus on a 'decoupled' version of the Haydys-Witten equations that appears in the context of knot invariants. I will discuss conditions under which the full equations simplify to the decoupled form. Since the latter exhibit a Hermitian Yang-Mills structure, these results may offer novel insights into the conjectured relationship to knot invariants.
                </p>
              </div>
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
        
          <!-- Include a card for each talk -->
          <div class="card-item col"><div class="card hoverable">
      <div class="row g-0">
<div class="col-md-12">
<div class="card-body">
            <!-- Title -->
            <h4 class="card-title">Topological Signatures of Convergence in Viral Evolution</h4>

            <!-- Venue and Date -->
            <p class="card-text">
              <strong>CompTopNN Meeting 2023, Sevilla</strong> <br>
              <i>November 08, 2023</i>
            </p>

            <!-- Optional link to slides --><p class="card-text"><a href="/assets/talks/2023_sevilla.pdf">Slides</a></p>
<!-- Optional abstract with expand/collapse --><div class="abstract">
                <a href="#" class="expand-link" onclick="toggleAbstract('abstract-9'); return false;">
                  Abstract
                </a>
                <p id="abstract-9" class="card-text abstract-text" style="display:none;">
                  The application of persistent homology to viral evolution was suggested some years ago at the example of genomic data of flu viruses. It was observed that long-lived topological features detect reassortment and recombination events, which provide much of the variability of flu viruses. Here, I instead present a careful investigation of short-lived features, typically interpreted as noise. Our results show that persistent homology efficiently identifies recurrent mutations in the evolution of SARS-CoV-2. This provides insights into how distinct viral strains exhibit convergent genomic structures, which is a hallmark of adaptation processes. Moreover, I will explain how our analysis can easily incorporate temporal information despite the general pitfalls of multi-persistent homology. In that way it is possible to monitor changes in convergent behaviour over the course of the pandemic.
                </p>
              </div>
</div>
        </div>
      </div>
    </div></div>

<script>
  function toggleAbstract(abstractId) {
    var abstract = document.getElementById(abstractId);
    if (abstract.style.display === "none") {
      abstract.style.display = "block";
    } else {
      abstract.style.display = "none";
    }
  }
</script>

        
      </div>
      </div>
    
  

</div>

  </article>
</div>

          </div>
        </div>
        
      
    </div>

    <!-- Footer -->    <footer class="sticky-bottom mt-5">
      <div class="container">Last updated: August 13, 2025.<br>
        Â© Copyright 2025 Michael  Bleher.<br><a href="https://subthaumic.github.io/impressum/">Impressum</a>.<i> Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </i>

      </div>
    </footer>

    <!-- JavaScripts -->
    <!-- jQuery -->
  <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>

    <!-- Bootsrap & MDB scripts -->
  <script src="/assets/js/bootstrap.bundle.min.js"></script>
  <!-- <script src="/assets/js/mdb.min.js"></script> -->
  <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script>

    <!-- Masonry & imagesLoaded -->
  <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script>
  <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@4/imagesloaded.pkgd.min.js"></script>
  <script defer src="/assets/js/masonry.js" type="text/javascript"></script>
    
  <!-- Medium Zoom JS -->
  <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script>
  <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script>
  <!-- Sidebar Table of Contents -->
  <script defer src="/assets/js/bootstrap-toc.min.js?c82ff4de8b0955d6ff14f5b05eed7eb6"></script>


  <!-- Bootstrap Table -->
  <script defer src="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.js"></script>

  <!-- Load Common JS -->
  <script src="/assets/js/no_defer.js?d633890033921b33e0ceb13d22340a9c"></script>
  <script defer src="/assets/js/common.js?acdb9690d7641b2f8d40529018c71a01"></script>
  <script defer src="/assets/js/copy_code.js?07b8786bab9b4abe90d10e61f7d12ff7" type="text/javascript"></script>

    

    <!-- MathJax -->
  <script type="text/javascript">
    window.MathJax = {
      tex: {
        tags: 'ams'
      }
    };
  </script>
  <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script>
  <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>

    
    

<!-- Scrolling Progress Bar -->
<script type="text/javascript">
  /*
   * This JavaScript code has been adapted from the article 
   * https://css-tricks.com/reading-position-indicator/ authored by Pankaj Parashar, 
   * published on the website https://css-tricks.com on the 7th of May, 2014.
   * Couple of changes were made to the original code to make it compatible 
   * with the `al-foio` theme.
   */
  const progressBar = $("#progress");
  /*
   * We set up the bar after all elements are done loading.
   * In some cases, if the images in the page are larger than the intended
   * size they'll have on the page, they'll be resized via CSS to accomodate
   * the desired size. This mistake, however, breaks the computations as the
   * scroll size is computed as soon as the elements finish loading.
   * To account for this, a minimal delay was introduced before computing the
   * values.
   */
  window.onload = function () {
    setTimeout(progressBarSetup, 50);
  };
  /*
   * We set up the bar according to the browser.
   * If the browser supports the progress element we use that.
   * Otherwise, we resize the bar thru CSS styling
   */
  function progressBarSetup() {
    if ("max" in document.createElement("progress")) {
      initializeProgressElement();
      $(document).on("scroll", function() {
        progressBar.attr({ value: getCurrentScrollPosition() });
      });
      $(window).on("resize", initializeProgressElement);
    } else {
      resizeProgressBar();
      $(document).on("scroll", resizeProgressBar);
      $(window).on("resize", resizeProgressBar);
    }
  }
  /*
   * The vertical scroll position is the same as the number of pixels that
   * are hidden from view above the scrollable area. Thus, a value > 0 is
   * how much the user has scrolled from the top
   */
  function getCurrentScrollPosition() {
    return $(window).scrollTop();
  }

  function initializeProgressElement() {
    let navbarHeight = $("#navbar").outerHeight(true);
    $("body").css({ "padding-top": navbarHeight });
    $("progress-container").css({ "padding-top": navbarHeight });
    progressBar.css({ top: navbarHeight });
    progressBar.attr({
      max: getDistanceToScroll(),
      value: getCurrentScrollPosition(),
    });
  }
  /*
   * The offset between the html document height and the browser viewport
   * height will be greater than zero if vertical scroll is possible.
   * This is the distance the user can scroll
   */
  function getDistanceToScroll() {
    return $(document).height() - $(window).height();
  }

  function resizeProgressBar() {
    progressBar.css({ width: getWidthPercentage() + "%" });
  }
  // The scroll ratio equals the percentage to resize the bar
  function getWidthPercentage() {
    return (getCurrentScrollPosition() / getDistanceToScroll()) * 100;
  }
</script>

    
  </body>
</html>
